### PR TITLE
add array support for request headers

### LINKED ISSUE #404: Client.dispatch headers argument as array
Allow passing an array as `headers` (i.e. the same format as the output from `dispatch({ onHeaders })`.

e.g.

```js
client.request({
 headers: ['key', 'value', 'key2', 'value2']
})
```

### PR DESCRIPTION
**Pull request by Ethan-Arrowood**
Closes #404

### CODE REVIEW

File: lib/core/request.js [hunk 1]
```diff
@@ -69,57 +69,70 @@ class Request {
 
     this.headers = ''
 
-    if (headers) {
-      for (const [key, val] of Object.entries(headers)) {
-        if (typeof val === 'object') {
-          throw new InvalidArgumentError(`invalid ${key} header`)
-        } else if (val === undefined) {
-          continue
+    const processHeader = (key, val) => {
```
Comment (ronag) [inline lib/core/request.js:72]: could you move this out as a global function?

File: lib/core/request.js [hunk 2]
```diff
@@ -69,57 +69,70 @@ class Request {
...
-          this.headers += `${key}: ${val}\r\n`
+    if (headers) {
+      if (Array.isArray(headers)) {
+        if (headers.length % 2 !== 0) {
+          throw new InvalidArgumentError('headers array must be even')
+        }
+        for (let i = 0; i < headers.length; i += 2) {
+          processHeader(headers[i], headers[i + 1])
+        }
+      } else {
```
Comment (ronag) [inline lib/core/request.js:133]: headers && typeof headers === 'object' else throw

Comment (Ethan-Arrowood) [reply lib/core/request.js:133]: this changes the API. Currently, the user doesn't have to specify any headers. Do you want me to change that?

Comment (Ethan-Arrowood) [reply lib/core/request.js:133]: Oh maybe you meant if headers are not undefined and are not of type object then throw error

**Review by mcollina — APPROVED** [review]
lgtm

File: lib/core/request.js [hunk 3]
```diff
@@ -69,66 +69,79 @@ class Request {
 
     this.headers = ''
 
-    if (headers) {
+    if (headers && Array.isArray(headers)) {
```
Comment (ronag) [inline lib/core/request.js:72]: You don't need the `headers &&` check here

File: lib/core/request.js [hunk 4]
```diff
@@ -69,66 +69,79 @@ class Request {
...
+    } else if (headers && typeof headers !== 'object') {
+      throw new InvalidArgumentError('headers must be an object or an array')
     }
 
     this[kRequestTimeout] = requestTimeout != null ? requestTimeout : 30e3
     this[kTimeout] = null
     this[kHandler] = handler
   }
 
+  processHeader (key, val) {
```
Comment (ronag) [inline lib/core/request.js:92]: Make private to file `processHeader(request, key, val)`

File: lib/core/request.js [hunk 5]
```diff
@@ -69,66 +69,79 @@ class Request {
...
-          key.length === 6 &&
-          key.toLowerCase() === 'expect'
-        ) {
-          throw new NotSupportedError('expect header not supported')
-        } else {
-          this.headers += `${key}: ${val}\r\n`
-        }
+        this.processHeader(key, val)
       }
+    } else if (headers && typeof headers !== 'object') {
```
Comment (ronag) [inline lib/core/request.js:83]:
```
else {
  throw
}
```

**Review by ronag — APPROVED** [review]
LGTM, good work

**Comment by ronag** [comment]
@Ethan-Arrowood could you also update the docs to reflect this change?

**Pull request merged**